x/data/types_test.go

```diff
@@ -0,0 +1,319 @@
+package data
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/require"
+)
+
+func TestContentHash_Validate(t *testing.T) {
+	tests := []struct {
+		name    string
+		sum     isContentHash_Sum
+		wantErr string
+	}{
+		{
+			"nil",
+			nil,
+			"invalid data.ContentHash type <nil>: unknown request",
+		},
+		{
+			"good raw",
+			&ContentHash_Raw_{Raw: &ContentHash_Raw{
+				Hash:            make([]byte, 32),
+				DigestAlgorithm: DigestAlgorithm_DIGEST_ALGORITHM_BLAKE2B_256,
+				MediaType:       MediaType_MEDIA_TYPE_UNSPECIFIED,
+			}},
+			"",
+		},
+		{
+			"good graph",
+			&ContentHash_Graph_{Graph: &ContentHash_Graph{
+				Hash:                      make([]byte, 32),
+				DigestAlgorithm:           DigestAlgorithm_DIGEST_ALGORITHM_BLAKE2B_256,
+				CanonicalizationAlgorithm: GraphCanonicalizationAlgorithm_GRAPH_CANONICALIZATION_ALGORITHM_URDNA2015,
+				MerkleTree:                GraphMerkleTree_GRAPH_MERKLE_TREE_NONE_UNSPECIFIED,
+			}},
+			"",
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			ch := ContentHash{
+				Sum: tt.sum,
+			}
+			err := ch.Validate()
+			if len(tt.wantErr) != 0 {
+				require.EqualError(t, err, tt.wantErr)
+			} else {
+				require.NoError(t, err)
+			}
+		})
+	}
+}
+
+func TestContentHash_Raw_Validate(t *testing.T) {
+	type fields struct {
+		Hash            []byte
+		DigestAlgorithm DigestAlgorithm
+		MediaType       MediaType
+	}
+	tests := []struct {
+		name    string
+		fields  fields
+		wantErr bool
+	}{
+		{
+			"good",
+			fields{
+				Hash:            make([]byte, 32),
+				DigestAlgorithm: DigestAlgorithm_DIGEST_ALGORITHM_BLAKE2B_256,
+				MediaType:       MediaType_MEDIA_TYPE_UNSPECIFIED,
+			},
+			false,
+		},
+		{
+			"bad",
+			fields{},
+			true,
+		},
+		{
+			"bad mediatype",
+			fields{
+				MediaType: -1,
+			},
+			true,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			chr := ContentHash_Raw{
+				Hash:            tt.fields.Hash,
+				DigestAlgorithm: tt.fields.DigestAlgorithm,
+				MediaType:       tt.fields.MediaType,
+			}
+			if err := chr.Validate(); (err != nil) != tt.wantErr {
+				t.Errorf("Validate() error = %v, wantErr %v", err, tt.wantErr)
```

Comment:
For my own understanding- what's the reason for using `t.Errorf()` here, but `require.Error()` / `require.NoError()` in other tests below?

Reply:
`t.Errorf` just fails the test with a formatted error message

`require.NoError` fails the test if and only if there is an error

Does that make sense?

Reply:
I think Cory question was more about an approach:
```
if wantErr {
   assert.Error(err)
} else {
   assert.NoError(err)
}
```
vs 
```
if (err != nil) != tt.wantErr {
   t.Errorf("Validate() error = %v, wantErr %v", err, tt.wantErr)
}
```

Reply:
Ah sorry...